pkg/controller/seed-controller-manager/cni-application-installation-controller/controller.go

```diff
@@ -0,0 +1,338 @@
+
+	cniPluginNamespace = "kube-system"
+)
+
+// UserClusterClientProvider provides functionality to get a user cluster client.
+type UserClusterClientProvider interface {
+	GetClient(ctx context.Context, c *kubermaticv1.Cluster, options ...clusterclient.ConfigOption) (ctrlruntimeclient.Client, error)
+}
+
+type Reconciler struct {
+	ctrlruntimeclient.Client
+
+	workerName                    string
+	recorder                      record.EventRecorder
+	userClusterConnectionProvider UserClusterClientProvider
+	log                           *zap.SugaredLogger
+	versions                      kubermatic.Versions
+	overwriteRegistry             string
+}
+
+func Add(ctx context.Context, mgr manager.Manager, numWorkers int, workerName string, userClusterConnectionProvider UserClusterClientProvider, log *zap.SugaredLogger, versions kubermatic.Versions, overwriteRegistry string) error {
+	reconciler := &Reconciler{
+		Client:                        mgr.GetClient(),
+		workerName:                    workerName,
+		recorder:                      mgr.GetEventRecorderFor(ControllerName),
+		userClusterConnectionProvider: userClusterConnectionProvider,
+		log:                           log.Named(ControllerName),
+		versions:                      versions,
+		overwriteRegistry:             overwriteRegistry,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	// Predicate for reacting to cluster update events only when CNIPlugin or ClusterNetwork config changed, or cluster Address changed
+	clusterUpdatePredicate := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			oldObj := e.ObjectOld.(*kubermaticv1.Cluster)
+			newObj := e.ObjectNew.(*kubermaticv1.Cluster)
+			if !reflect.DeepEqual(oldObj.Spec.CNIPlugin, newObj.Spec.CNIPlugin) {
+				return true
+			}
+			if !reflect.DeepEqual(oldObj.Spec.ClusterNetwork, newObj.Spec.ClusterNetwork) {
+				return true
+			}
+			if !reflect.DeepEqual(oldObj.Status.Address, newObj.Status.Address) {
+				return true
+			}
+			return false
+		},
+	}
+
+	// Watch on cluster events
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, &handler.EnqueueRequestForObject{}, clusterUpdatePredicate, workerlabel.Predicates(workerName)); err != nil {
+		return fmt.Errorf("failed to create watch: %w", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := r.log.With("cluster", request.NamespacedName.Name)
+	log.Debug("Processing")
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, request.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if cluster.DeletionTimestamp != nil {
```

Comment:
### General question:
In our codebase, to check if an object is in deletion state, I often see 2 different versions.

-  testing DeletionTimestamp is not nill 
-  testing DeletionTimestamp is not zero.
    ```
    if !cluster.DeletionTimestamp.IsZero() {
		    // Cluster is queued for deletion; no action required
		    log.Debugw("Cluster is queued for deletion; skipping")
		    return reconcile.Result{}, nil
	    }
    ```

I was wondering, what was the best practice? 

Additionally to test if the object is nil, `IsZero()` function also checks is time is zero.  To me, It seems highly improbable to have a DeletionTimestamp not nil and not initialized, isn't it?




In Kubebuilder book, I see only one example with DeletionTimestamp and they use the IsZero function for the test...

In the end, I don't think it really matters but I'm curious...

Reply:
Probably doesn't matter.  

But having a mix might lead to bad scenarios. 

Say we have mix of nil checks and IsZero. 

Code that is using `IsZero` can be fooled into 
thinking that a resource is not marked for deletion by setting value of `DeletionTimestamp` to 0
while code that is using nil checks thinks the resource is marked for deletion. 

I vote for nil check everywhere. 

I could for  IsZero but we are not supposed 
to use the value of the timestamp just that it is set
and IsZero breaks the rule by looking at the value.  
Also, requires knowing that such a function is there.

Reply:
Frankly, I don't have a preference on this :)

Reply:
Delete is now filtered by predicate so this question is no longer important ;) anyway I would still keep the check here just in case - e.g. if predicate would change in the future.